File: thien_tam_app/lib/features/audio/data/models/audio.dart
class Audio {
  final String id;
  final String title;
  final String? description;
  final String? artist;
  final int? duration; // in seconds
  final String category;
  final List<String> tags;
  final String cloudinaryPublicId;
  final String cloudinaryUrl;
  final String cloudinarySecureUrl;
  final int? fileSize;
  final String? format;
  final int playCount;
  final bool isPublic;
  final DateTime createdAt;
  final DateTime updatedAt;

  Audio({
    required this.id,
    required this.title,
    this.description,
    this.artist,
    this.duration,
    required this.category,
    required this.tags,
    required this.cloudinaryPublicId,
    required this.cloudinaryUrl,
    required this.cloudinarySecureUrl,
    this.fileSize,
    this.format,
    required this.playCount,
    required this.isPublic,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Audio.fromJson(Map<String, dynamic> json) {
    return Audio(
      id: json['_id'] as String,
      title: json['title'] as String,
      description: json['description'] as String?,
      artist: json['artist'] as String?,
      duration: json['duration']?.toInt(),
      category: json['category'] as String,
      tags:
          (json['tags'] as List<dynamic>?)?.map((e) => e as String).toList() ??
          [],
      cloudinaryPublicId: json['cloudinaryPublicId'] as String,
      cloudinaryUrl: json['cloudinaryUrl'] as String,
      cloudinarySecureUrl: json['cloudinarySecureUrl'] as String,
      fileSize: json['fileSize'] as int?,
      format: json['format'] as String?,
      playCount: json['playCount'] as int? ?? 0,
      isPublic: json['isPublic'] as bool? ?? true,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
    );
  }

  String get streamUrl => cloudinarySecureUrl;

  String get durationFormatted {
    if (duration == null) return '--:--';
    final minutes = duration! ~/ 60;
    final seconds = duration! % 60;
    return '${minutes.toString().padLeft(2, '0')}:${seconds.toString().padLeft(2, '0')}';
  }

  String get fileSizeFormatted {
    if (fileSize == null) return '--';
    if (fileSize! < 1024) return '${fileSize}B';
    if (fileSize! < 1024 * 1024)
      return '${(fileSize! / 1024).toStringAsFixed(1)}KB';
    return '${(fileSize! / (1024 * 1024)).toStringAsFixed(1)}MB';
  }
}

class AudioCategory {
  final String value;
  final String label;
  final String description;

  AudioCategory({
    required this.value,
    required this.label,
    required this.description,
  });

  factory AudioCategory.fromJson(Map<String, dynamic> json) {
    return AudioCategory(
      value: json['value'] as String,
      label: json['label'] as String,
      description: json['description'] as String,
    );
  }
}

File: backend_thien_tam_app/src/models/Audio.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface IAudio extends Document {
  title: string;
  description?: string;
  artist?: string;
  duration?: number; // in seconds
  category: 'sutra' | 'mantra' | 'dharma-talk' | 'meditation' | 'chanting' | 'music' | 'other';
  tags: string[];
  cloudinaryPublicId: string;
  cloudinaryUrl: string;
  cloudinarySecureUrl: string;
  fileSize?: number; // in bytes
  format?: string; // mp3, wav, etc.
  uploadedBy: mongoose.Types.ObjectId;
  playCount: number;
  isPublic: boolean;
  createdAt: Date;
  updatedAt: Date;
}

const AudioSchema: Schema = new Schema(
  {
    title: {
      type: String,
      required: [true, 'Title is required'],
      trim: true,
      maxlength: [200, 'Title cannot exceed 200 characters'],
    },
    description: {
      type: String,
      trim: true,
      maxlength: [1000, 'Description cannot exceed 1000 characters'],
    },
    artist: {
      type: String,
      trim: true,
      maxlength: [100, 'Artist name cannot exceed 100 characters'],
    },
    duration: {
      type: Number,
      min: [0, 'Duration cannot be negative'],
    },
    category: {
      type: String,
      enum: {
        values: ['sutra', 'mantra', 'dharma-talk', 'meditation', 'chanting', 'music', 'other'],
        message: '{VALUE} is not a valid category',
      },
      required: [true, 'Category is required'],
      default: 'other',
    },
    tags: {
      type: [String],
      default: [],
      validate: {
        validator: function (tags: string[]) {
          return tags.length <= 10;
        },
        message: 'Cannot have more than 10 tags',
      },
    },
    cloudinaryPublicId: {
      type: String,
      required: [true, 'Cloudinary public ID is required'],
      unique: true,
    },
    cloudinaryUrl: {
      type: String,
      required: [true, 'Cloudinary URL is required'],
    },
    cloudinarySecureUrl: {
      type: String,
      required: [true, 'Cloudinary secure URL is required'],
    },
    fileSize: {
      type: Number,
      min: [0, 'File size cannot be negative'],
    },
    format: {
      type: String,
      lowercase: true,
    },
    uploadedBy: {
      type: Schema.Types.ObjectId,
      ref: 'AdminUser',
      required: [true, 'Uploader is required'],
    },
    playCount: {
      type: Number,
      default: 0,
      min: [0, 'Play count cannot be negative'],
    },
    isPublic: {
      type: Boolean,
      default: true,
    },
  },
  {
    timestamps: true,
  }
);

// Indexes for better query performance
AudioSchema.index({ category: 1, isPublic: 1 });
AudioSchema.index({ tags: 1 });
AudioSchema.index({ createdAt: -1 });
AudioSchema.index({ playCount: -1 });
AudioSchema.index({ title: 'text', description: 'text', artist: 'text' });

// Method to increment play count
AudioSchema.methods.incrementPlayCount = async function () {
  this.playCount += 1;
  return this.save();
};

export default mongoose.model<IAudio>('Audio', AudioSchema);

